Unable to build on wasm32-unknown-unknown due to libc malloc

Creating a project with lodepng as a dependency and building with `--target wasm32-unknown-unknown` yields the following errors:

```
   Compiling lodepng v2.5.0
error[E0425]: cannot find function `malloc` in crate `libc`
  --> /home/carado/.cargo/registry/src/github.com-1ecc6299db9ec823/lodepng-2.5.0/src/rustimpl.rs:53:15
   |
53 |         libc::malloc(size) as *mut _
   |               ^^^^^^ not found in `libc`

error[E0425]: cannot find function `realloc` in crate `libc`
  --> /home/carado/.cargo/registry/src/github.com-1ecc6299db9ec823/lodepng-2.5.0/src/rustimpl.rs:58:11
   |
58 |     libc::realloc(ptr as *mut _, size) as *mut _
   |           ^^^^^^^ not found in `libc`
   |
help: possible candidate is found in another module, you can import it into scope
   |
16 | use std::alloc::realloc;
   |

error[E0425]: cannot find function `free` in crate `libc`
  --> /home/carado/.cargo/registry/src/github.com-1ecc6299db9ec823/lodepng-2.5.0/src/rustimpl.rs:62:11
   |
62 |     libc::free(ptr as *mut _)
   |           ^^^^ not found in `libc`

error[E0425]: cannot find function `free` in crate `libc`
   --> /home/carado/.cargo/registry/src/github.com-1ecc6299db9ec823/lodepng-2.5.0/src/lib.rs:818:11
    |
818 |     libc::free(ptr as *mut _);
    |           ^^^^ not found in `libc`

error: aborting due to 4 previous errors
```
Is using the system's malloc and free really necessary ?

--- a/src/ffi.rs
+++ b/src/ffi.rs
@@ -8,10 +8,17 @@ use std::slice;
 use std::fmt;
 use std::os::raw::{c_char, c_uint, c_long, c_void};
 use std::ffi::CStr;
-use std::path::*;
 use std::io;
 
 use crate::rustimpl;
+#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+mod libc;
+
+#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+use std::path::{PathBuf};
+
+#[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
+use std::path::{Path};
 
 macro_rules! lode_error {
     ($e:expr) => {

--- a/src/ffi.rs
+++ b/src/ffi.rs
@@ -8,17 +8,12 @@ use std::slice;
 use std::fmt;
 use std::os::raw::{c_char, c_uint, c_long, c_void};
 use std::ffi::CStr;
+use std::path::*;
 use std::io;
 
 use crate::rustimpl;
 #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
 mod libc;
-
-#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
-use std::path::PathBuf;
-
-#[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
-use std::path::Path;
 
 macro_rules! lode_error {
     ($e:expr) => {

--- a/src/ffi.rs
+++ b/src/ffi.rs
@@ -15,10 +15,10 @@ use crate::rustimpl;
 mod libc;
 
 #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
-use std::path::{PathBuf};
+use std::path::PathBuf;
 
 #[cfg(not(all(target_arch = "wasm32", target_os = "unknown")))]
-use std::path::{Path};
+use std::path::Path;
 
 macro_rules! lode_error {
     ($e:expr) => {

--- a/src/ffi/libc.rs
+++ b/src/ffi/libc.rs
@@ -1,0 +1,36 @@
+pub type size_t = usize;
+
+const MALLOC_HEADER : isize = 8;
+const MALLOC_ALIGN : usize = 8;
+
+use super::c_void;
+use std::alloc::{self, Layout};
+use std::ptr;
+
+pub unsafe fn malloc(size: size_t) -> *mut c_void {
+    let lay = Layout::from_size_align_unchecked(MALLOC_HEADER as usize + size, MALLOC_ALIGN);
+    let p = alloc::alloc(lay);
+    if p.is_null() {
+        return ptr::null_mut();
+    }
+    *(p as *mut size_t) = size;
+    p.offset(MALLOC_HEADER) as *mut c_void
+}
+pub unsafe fn free(p: *mut c_void) {
+    let p = p.offset(-MALLOC_HEADER) as *mut u8;
+    let size = *(p as *mut size_t);
+    let lay = Layout::from_size_align_unchecked(MALLOC_HEADER as usize + size, MALLOC_ALIGN);
+    alloc::dealloc(p, lay);
+}
+pub unsafe fn realloc(p: *mut c_void, _size: size_t) -> *mut c_void {
+    let p = p.offset(-MALLOC_HEADER) as *mut u8;
+    let size = *(p as *mut size_t);
+    let lay = Layout::from_size_align_unchecked(MALLOC_HEADER as usize + size, MALLOC_ALIGN);
+    let p = alloc::realloc(p, lay, size);
+    if p.is_null() {
+        return ptr::null_mut();
+    }
+    *(p as *mut size_t) = size;
+    p.offset(MALLOC_HEADER) as *mut c_void
+}
+

